MINOR: replace test "expected" parameter by assertThrows

chia7712 opened the pull request:
`assertThrows` can make checks in testing more flexible and accurate so I think it is worth substituting `assertThrows` for "expected" parameter.

This PR includes following changes.

1. ```@Test(expected = Exception.class)``` is replaced by ```assertThrows```
1. remove reference to ```org.scalatest.Assertions``` 
1. change the magic code from 1 to 2 for ```testAppendAtInvalidOffset``` to test ZSTD
1. rename ```testMaybeAddPartitionToTransactionXXXX``` to ```testNotReadyForSendXXX```
1. increase maxBlockMs from 1s to 3s to avoid unexpected timeout from ```TransactionsTest#testTimeout``` 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 committed:
MINOR: replace test "expected" parameter by assertThrows

ijuma commented on clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java:
```diff
@@ -117,13 +118,12 @@ public void setup() {
         long now = time.milliseconds();
         ClientRequest request = client.newClientRequest("5", builder, now, false);
-        client.send(request, now);
-        client.poll(1, time.milliseconds());
```
We don't need this anymore?

chia7712 replied:
```client.send(request, now);``` throws exception so ```client.poll(1, time.milliseconds());``` was never reached.

chia7712 commented:
@ijuma Any thoughts to this PR?

ijuma commented on clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:
```diff
@@ -196,12 +197,12 @@ public void testHierarchicalSensors() {
         Sensor c1 = metrics.sensor("child1", p);
         Sensor c2 = metrics.sensor("child2", p);
-        metrics.sensor("gc", c1, c2); // should fail
+        assertThrows(IllegalArgumentException.class, () -> metrics.sensor("gc", c1, c2)); // should fail
```
We can remove the comment, it's redundant.

ijuma commented on clients/src/test/java/org/apache/kafka/common/record/ByteBufferLogInputStreamTest.java:
```diff
@@ -116,8 +117,7 @@ public void iteratorRaisesOnTooLargeRecords() {
         ByteBufferLogInputStream logInputStream = new ByteBufferLogInputStream(buffer, 25);
-        assertNotNull(logInputStream.nextBatch());
-        logInputStream.nextBatch();
+        assertThrows(CorruptRecordException.class, logInputStream::nextBatch);
```
Hmm, I think this was intended to fail on the second `nextBatch`, so maybe we should adjust the test.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -411,14 +411,13 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
     consumer.assign(List(tp).asJava)
     // the exception is expected when the consumer attempts to lookup offsets
-    consumeRecords(consumer)
-    confirmReauthenticationMetrics()
```
We should not remove this, right?

chia7712 replied:
good catching!

ijuma reviewed (commented):
Thanks for the PR. Looks good overall, a few minor comments below.

One more thing: we seem to sometimes use the scalatest asserts and sometimes the JUnit ones for the Scala code. I would suggest using the JUnit ones consistently as it will make it easier to convert the tests to JUnit 5.

chia7712 committed:
replace org.scalatest.Assertions by junit as much as possible

chia7712 commented:
> One more thing: we seem to sometimes use the scalatest asserts and sometimes the JUnit ones for the Scala code. I would suggest using the JUnit ones consistently as it will make it easier to convert the tests to JUnit 5.

Except for ```org.scalatest.Assertions.withClue```, others package from ```org.scalatest.Assertions``` are replaced by Junit.

chia7712 commented:
@ijuma the latest commit replaces all ExpectedException or @Test(expected = Exception.class) by assertThrows. It seems to me we can get closer to junit 5 if this PR is got into trunk. Furthermore, I'd like to file following PR to remove all junit 4 code.

1. Replace ScalaTest by junit
2. Replace junit 4 APIs by Junit 5 for each module
    1. rewrite parameter tests
    2. replace ```org.junit``` by ```org.junit.jupiter.api```
    3. replace ```IntegrationTest``` by tag ```Integration```
    4. replace ```useJUnitPlatform``` by ```useJUnit``` 

WDYT?

chia7712 committed:
replace all ExpectedException or @test(expected = Exception.class) by assertThrows

ijuma commented:
@chia7712 Thanks. I was about halfway through the review of the second to last commit. :) Let's get this PR over the line and not change it further please. Yes, I had filed a few JIRAs for migrating to JUnit 5, see https://issues.apache.org/jira/browse/KAFKA-7339.

I think we should do it on a module per module basis to make it easier to review (the JIRAs are structured in that way). I was going to try the clients module after this PR is merged. Feel free to take one of the other ones. `core` would be a good candidate perhaps.

chia7712 commented:
> I was about halfway through the review of the second to last commit. :) Let's get this PR over the line and not change it further please. 

I am so sorry my force push obstructed you from reviewing. I will never do that again :(

chia7712 commented:
> Feel free to take one or more of the other ones. core would be a good candidate perhaps.

I’d like to take over both core module and stream module :)

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -192,11 +192,12 @@ public void testConstructorWithSerializers() {
         Properties producerProps = new Properties();
         producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000");
-        new KafkaProducer(producerProps);
+
```
Was this new line intentional?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:
```diff
@@ -663,12 +623,6 @@ public void shouldPreserveOffsetsFromCommitByGroupMetadataOnAbortIfTransactionsA
-                put(new TopicPartition(topic, 3), new OffsetAndMetadata(84L, null));
-            }
-        };
         producer.sendOffsetsToTransaction(groupCommit, new ConsumerGroupMetadata(group2));
```
I think this was intended to be `groupCommit2`, right?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
     public void testMaybeAddPartitionToTransactionBeforeInitTransactions() {
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(IllegalStateException.class, () -> transactionManager.failIfNotReadyForSend());
```
This change doesn't seem right. The test is intended to test `maybeAddPartitionToTransaction`.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
         doInitTransactions();
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(IllegalStateException.class, () -> transactionManager.failIfNotReadyForSend());
```
Same as above.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
         transactionManager.transitionToAbortableError(new KafkaException());
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(KafkaException.class, () -> transactionManager.failIfNotReadyForSend());
```
Same as above.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
         transactionManager.transitionToFatalError(new KafkaException());
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(KafkaException.class, () -> transactionManager.failIfNotReadyForSend());
```
Same as above.

chia7712 replied:
https://github.com/apache/kafka/commit/717c55be971df862c55f55d245b9997f1d6f998c moved the check of state out of ```maybeAddPartitionToTransaction``` and it added ```failIfNotReadyForSend``` to all test cases to make them throw exception. I will update the test cases name to avoid confusion.

ijuma reviewed (commented):
I have reviewed about 60 files and left a few comments. It would be good to clearly mention the cases where the translation wasn't mechanism as they are the most likely sources of issues.

ijuma commented on clients/src/test/java/org/apache/kafka/common/memory/GarbageCollectedMemoryPoolTest.java:
```diff
@@ -23,68 +23,74 @@
         GarbageCollectedMemoryPool pool = new GarbageCollectedMemoryPool(1000, 10, true, null);
         ByteBuffer fellOffATruck = ByteBuffer.allocate(1);
-        pool.release(fellOffATruck);
-        pool.close();
```
Do we still need to call `close` to avoid a leak?

ijuma commented on clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:
```diff
@@ -504,20 +510,21 @@ public void writePastLimit() {
 
         long logAppendTime = System.currentTimeMillis();
-        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, RecordBatch.MAGIC_VALUE_V1, compressionType,
+        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, RecordBatch.MAGIC_VALUE_V2, compressionType,
```
Why did we change the magic value version here?

chia7712 replied:
magic 1 does not support ```ZStandard```

ijuma replied:
Unrelated to the changes in this PR then, just something you noticed? Worth a line in the PR description to avoid confusion.

chia7712 replied:
will copy that

ijuma commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticatorTest.java:
```diff
@@ -67,8 +68,7 @@ public void testOversizeRequest() throws IOException {
             return 4;
         });
-        authenticator.authenticate();
-        verify(transportLayer).read(any(ByteBuffer.class));
```
Why don't we need the `verify` call?

ijuma commented on connect/api/src/test/java/org/apache/kafka/connect/data/SchemaBuilderTest.java:
```diff
@@ -293,16 +295,13 @@ public void testEmptyStruct() {
+            .name("testing")
+            .field("id", SchemaBuilder.string().doc("").build())
+            .field("id", SchemaBuilder.string().doc("").build())
+            .build());
```
It seems that we are not testing the `validate` method anymore.

chia7712 replied:
```StructTest``` has test cases for ```validate``` method.

chia7712 committed:
address review comment

chia7712 commented:
@ijuma Thanks for your great reviewing. I have addressed most review comment in latest commit.

> It would be good to clearly mention the cases where the translation wasn't mechanism as they are the most likely sources of issues.

Pardon me. I failed to catch your point.

chia7712 committed:
fix conflicts

chia7712 commented:
rebase code to fix conflicts

ijuma commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:
```diff
@@ -49,18 +49,17 @@
     public void shouldThrowStreamsExceptionIfExceptionCaughtDuringClose() {
         final ProcessorNode node = new ProcessorNode("name", new ExceptionalProcessor(), Collections.emptySet());
-        node.init(null);
-        node.close();
```
Should we keep the `close`?

chia7712 replied:
It fails to init so the close is unnecessary.

ijuma replied:
I don't think this is necessarily the case. The failure during init can happen after some resources have been allocated, right?

ijuma commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -715,7 +711,7 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
   private def cleanMetricsRegistry(): Unit = {
     val metrics = KafkaYammerMetrics.defaultRegistry
-    metrics.allMetrics.keySet.forEach(metrics.removeMetric)
+    metrics.allMetrics.keySet.forEach(m => metrics.removeMetric(m))
```
Why did we change this?

chia7712 replied:
IntelliJ can't be aware that lambda function :(

Will revert it.

ijuma commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -407,46 +406,40 @@ class TransactionsTest extends KafkaServerTestHarness {
                   timeoutProcess: KafkaProducer[Array[Byte], Array[Byte]] => Unit): Unit = {
-    val producer = createTransactionalProducer("transactionProducer", maxBlockMs =  1000)
-
+    val producer = createTransactionalProducer("transactionProducer", maxBlockMs = 3000)
```
Why did we increase maxBlockMs here?

chia7712 replied:
```1000``` is too small to complete ```createTransactionalProducer``` and it causes ```TimeoutException```. I don't think those test cases tried to test timeout on ```createTransactionalProducer```.

ijuma commented on core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala:
```diff
@@ -17,24 +17,21 @@
   private val sampleWindowSec = 1
   private val maxThrottleTimeMs = 500
-  private val metricName = new MetricName("test-metric", "groupA", "testA", Map.empty.asJava)
+  private val metricName = new MetricName("test-metric", "groupA", "testA", java.util.Collections.emptyMap())
```
Why did you change this?

chia7712 replied:
my bad and will revert it. (IntelliJ complained about the erased types)

ijuma replied:
I would ignore IntelliJ issues if it compiles without warnings. I think there are a few more changes like this in the middle of thousands of lines of diff. Something to avoid in the future. :)

chia7712 replied:
Please give me one second to revert those changes :(

ijuma reviewed (commented):
Reviewed a bit over 200 files and a few questions/comments.

ijuma commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -343,7 +340,7 @@ class DynamicBrokerConfigTest {
     class TestAuthorizer extends Authorizer with Reconfigurable {
       @volatile var superUsers = ""
-      override def start(serverInfo: AuthorizerServerInfo): util.Map[Endpoint, _ <: CompletionStage[Void]] = Map.empty.asJava
+      override def start(serverInfo: AuthorizerServerInfo): util.Map[Endpoint, _ <: CompletionStage[Void]] = util.Collections.emptyMap()
```
Why did we change this?

chia7712 committed:
revert unnecessary changes

chia7712 committed:
revert changes

chia7712 committed:
revert more changes

chia7712 commented:
I have reverted most unrelated changes.  see https://github.com/apache/kafka/pull/9520/commits/d0e2236e96db0b514f4d2c07641eb236feb1de74

ijuma commented:
@chia7712 Can you please remove the last commit? It seems to have a mix of reverts and it's triggering changes to files I had already reviewed.

chia7712 committed:
Revert "revert more changes"

This reverts commit d0e2236e96db0b514f4d2c07641eb236feb1de74.

chia7712 commented:
> Can you please remove the last commit?

done!

ijuma reviewed (approved):
Thanks for the PR, LGTM. Will you do the `core` conversion to JUnit 5 next? We are looking into how to convert some of the tests to work in both raft and zk modes and it would be easier to do that with JUnit 5.

ijuma commented:
Looking at the test results, it looks like there are some real failures. For example:

`Build / JDK 8 / kafka.utils.QuotaUtilsTest.testBoundedThrottleTimeThrowsExceptionIfProvidedNonRateMetric`

chia7712 commented:
> Will you do the core conversion to JUnit 5 next?

of course. Should we keep using ```ScalaTest``` and ```hamcrest```? They are not in conflict with Junit 5 but it results in various code style in testing.

chia7712 committed:
fix QuotaUtilsTest

ijuma commented:
I would do the minimal set of changes to make the test suite pass with JUnit 5.

chia7712 commented:
fix conflicting files

chia7712 commented:
@ijuma Thanks for all your great reviews!
